hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/clean/CleanPlanner.java

```diff
@@ -368,16 +377,23 @@ public CleanPlanner(HoodieEngineContext context, HoodieTable<T, I, K, O> hoodieT
   
   private List<CleanFileInfo> getReplacedFilesEligibleToClean(List<String> savepointedFiles, String partitionPath, Option<HoodieInstant> earliestCommitToRetain) {
     final Stream<HoodieFileGroup> replacedGroups;
-    if (earliestCommitToRetain.isPresent()) {
+    if (earliestCommitToRetain.isPresent() && !isDeletePartitionOperation(earliestCommitToRetain)) {
       replacedGroups = fileSystemView.getReplacedFileGroupsBefore(earliestCommitToRetain.get().getTimestamp(), partitionPath);
     } else {
       replacedGroups = fileSystemView.getAllReplacedFileGroups(partitionPath);
     }
-    return replacedGroups.flatMap(HoodieFileGroup::getAllFileSlices)
+
+    Stream<CleanFileInfo> replacedGroupsStream = replacedGroups.flatMap(HoodieFileGroup::getAllFileSlices)
         // do not delete savepointed files  (archival will make sure corresponding replacecommit file is not deleted)
         .filter(slice -> !slice.getBaseFile().isPresent() || !savepointedFiles.contains(slice.getBaseFile().get().getFileName()))
-        .flatMap(slice -> getCleanFileInfoForSlice(slice).stream())
-        .collect(Collectors.toList());
+        .flatMap(slice -> getCleanFileInfoForSlice(slice).stream());
+
+    Stream<CleanFileInfo> dropPartitionsStream = getDropPartitionPathsForClean(earliestCommitToRetain)
```

Comment:
we can do something like this to get the list of partitions to be deleted w/o relying on earliest instant to retain. 

at L 385, 

we can call getAllFileGroups() and filter out replacedGroups(collected in lines 380 to 384). If resultant set if empty, we know we need to clean up entire partition and hence add partition path to the list of paths to be cleaned up.